[SelectionDAG] Add PARTIAL_REDUCE_U/SMLA ISD Nodes

JamesChesterman opened the pull request:
Add signed and unsigned PARTIAL_REDUCE_MLA ISD nodes. Add command line argument (new-partial-reduce-lowering) that indicates whether the intrinsic experimental_vector_partial_ reduce_add will be transformed into the new ISD node. Lowering with the new ISD nodes will, for now, always be done as an expand.

JamesChesterman committed:
[AArch64] Add PARTIAL_REDUCE_U/SMLA ISD Nodes

Add signed and unsigned PARTIAL_REDUCE_MLA ISD nodes.
Add command line argument (new-partial-reduce-lowering) that
indicates whether the intrinsic experimental_vector_partial_
reduce_add will be transformed into the new ISD node.
Lowering with the new ISD nodes will, for now, always be done as
an expand.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -2473,6 +2473,23 @@ SDValue SelectionDAG::getShiftAmountOperand(EVT LHSTy, SDValue Op) {
+
+  EVT FullTy = Input1.getValueType();
+
+  SDValue Input = Input1;
+  APInt ConstantOne;
+  if (!ISD::isConstantSplatVector(Input2.getNode(), ConstantOne) ||
+      !ConstantOne.isOne())
+    Input = getNode(ISD::MUL, DL, FullTy, Input1, Input2);
```
This code should extend `Input1` and `Input2` to the element type of `Acc`.

JamesChesterman replied:
Done

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2824,6 +2833,12 @@ SDValue DAGTypeLegalizer::PromoteIntOp_VECTOR_FIND_LAST_ACTIVE(SDNode *N,
 }
 
+SDValue DAGTypeLegalizer::PromoteIntOp_PARTIAL_REDUCE_MLA(SDNode *N) {
+  SDValue Res = DAG.expandPartialReduceMLA(N);
+  ReplaceValueWith(SDValue(N, 0), Res);
+  return SDValue();
```
This code should just promote the operands to a legal type.

This will however mean that if all types are legal, that the target will not be able to select the operation. This you can fix by marking the operations as `Expand` for AArch64, and calling the expand code from `LegalizeVectorOps`. Any sign/zero-extension of the operands by the expand code, will then be revisited by a second type-legalization stage.

JamesChesterman replied:
Done

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -135,6 +135,10 @@ static cl::opt<unsigned> SwitchPeelThreshold(
              "switch statement. A value greater than 100 will void this "
              "optimization"));
 
+static cl::opt<bool> NewPartialReduceLowering(
```
nit: maybe add a FIXME to say that this flag is temporary, to help transition to do proper lowering with the new selectiondag nodes?

JamesChesterman replied:
Done

JamesChesterman committed:
Address comments on PR.

JamesChesterman committed:
Address comments

Move the getPartialReduceAdd function around.
Make the new codepath work for fixed length NEON vectors too.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -6139,6 +6154,12 @@ SDValue DAGTypeLegalizer::PromoteIntRes_VECTOR_FIND_LAST_ACTIVE(SDNode *N) {
 }
 
+SDValue DAGTypeLegalizer::PromoteIntRes_PARTIAL_REDUCE_MLA(SDNode *N) {
+  SDValue Res = DAG.expandPartialReduceMLA(N);
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -2473,6 +2473,23 @@ SDValue SelectionDAG::getShiftAmountOperand(EVT LHSTy, SDValue Op) {
   return getZExtOrTrunc(Op, SDLoc(Op), ShTy);
 }
 
+SDValue SelectionDAG::expandPartialReduceMLA(SDNode *N) {
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -12114,3 +12114,41 @@ SDValue TargetLowering::expandVectorNaryOpBySplitting(SDNode *Node,
+    Input = ExtendToAccEltVT(Input1);
+  }
+
+  return DAG.getPartialReduceAdd(DL, ReducedTy, Acc, Input);
```
Done. I've then called `TargetLowering::getPartialReduceAdd()` in the `performIntrinsicCombine()` function so no new behaviour is performed.

JamesChesterman committed:
Combine two functions together

JamesChesterman commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -5624,7 +5624,14 @@ class TargetLowering : public TargetLoweringBase {
+  // its operands and ReducedTY is the return type.
+  static SDValue getPartialReduceAdd(SDLoc DL, EVT ReducedTy, EVT FullTy,
+                                     SDValue Op1, SDValue Op2,
+                                     SelectionDAG &DAG);
```
Done, sorry I misunderstood, I thought we might have wanted to keep the two functions separate for now and have the new code path follow the `expandPartialReduceMLA()` function. But I understand now that they would do the same thing, so I've just combined the two functions into the `expandPartialReduceMLA()` function.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -12114,3 +12115,65 @@ SDValue TargetLowering::expandVectorNaryOpBySplitting(SDNode *Node,
+    Input = DAG.getNode(ISD::MUL, DL, NewVT, Input1, Input2);
+  } else {
+    Input = ExtendToAccEltVT(Input1);
+  }
```
Done

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -12188,3 +12189,51 @@ SDValue TargetLowering::scalarizeExtractedVectorLoad(EVT ResultVT,
   return Load;
 }
+
+SDValue TargetLowering::expandPartialReduceMLA(SDLoc DL, SDValue Acc,
+                                               SDValue Input1, SDValue Input2,
+                                               SelectionDAG &DAG) const {
```
Could you change this so that `expandPartialReduceMLA` only takes a `Node *` and a `DAG`? That's what the other lowering functions seem to do.

That does mean rejigging the code in SelectionDAGBuilder a little bit to return a newly created node, and there is a function in `performIntrinsicCombine` as well that would need updating to take a `PARTIAL_REDUCE_UMLA` node.

JamesChesterman replied:
Done

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/neon-partial-reduce-dot-product.ll:
```diff
@@ -2,6 +2,7 @@
 ; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod < %s | FileCheck %s --check-prefixes=CHECK,CHECK-DOT,CHECK-NOI8MM
 ; RUN: llc -mtriple aarch64 -mattr=+neon < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NOI8MM,CHECK-NODOT
 ; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod,+i8mm < %s | FileCheck %s --check-prefixes=CHECK,CHECK-DOT,CHECK-I8MM
+; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod,+i8mm -new-partial-reduce-lowering < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NEWLOWERING
```
With the suggestion below, you don't need the new check lines.
```suggestion
; RUN: llc -mtriple aarch64 -mattr=+neon,+dotprod,+i8mm -new-partial-reduce-lowering < %s | FileCheck %s --check-prefixes=CHECK,CHECK-NOI8MM,CHECK-NODOT
```

JamesChesterman replied:
Done

sdesmalen-arm commented on llvm/test/CodeGen/AArch64/sve-partial-reduce-dot-product.ll:
```diff
@@ -1,12 +1,41 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
 ; RUN: llc -mtriple=aarch64 -mattr=+sve2,+i8mm %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-I8MM
 ; RUN: llc -mtriple=aarch64 -mattr=+sve2 %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-NOI8MM
+; RUN: llc -mtriple=aarch64 -mattr=+sve2,+i8mm -new-partial-reduce-lowering %s -o - | FileCheck %s --check-prefixes=CHECK,CHECK-NEWLOWERING
```
By removing `CHECK` from this line, we only add check lines for the NEWLOWERING case, and avoid duplicate lines for `CHECK-I8MM` and `CHECK-NOI8MM`.

```suggestion
; RUN: llc -mtriple=aarch64 -mattr=+sve2,+i8mm -new-partial-reduce-lowering %s -o - | FileCheck %s --check-prefixes=CHECK-NEWLOWERING
```

JamesChesterman replied:
Done

sdesmalen-arm reviewed (commented):
Thanks, this is looking pretty good! Just a few more minor comments before I'm happy to accept it.

JamesChesterman committed:
Address comments on PR. Involves changing arguments on the
function.

JamesChesterman commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1451,6 +1451,19 @@ enum NodeType {
   VECREDUCE_UMIN,
 
+  // PARTIAL_REDUCE_[U|S]MLA(Accumulator, Input1, Input2)
+  // The partial reduction nodes sign-or zero extend Input1 and Input2 to
```
Done

JamesChesterman commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -5634,6 +5634,11 @@ class TargetLowering : public TargetLoweringBase {
+  // consisting of zext/sext, extract_subvector, mul and add operations.
+  SDValue expandPartialReduceMLA(SDLoc DL, SDValue Acc, SDValue Input1,
+                                 SDValue Input2, SelectionDAG &DAG) const;
+
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -6196,6 +6212,16 @@ SDValue DAGTypeLegalizer::PromoteIntRes_VECTOR_FIND_LAST_ACTIVE(SDNode *N) {
+  SDValue ExtAcc = GetPromotedInteger(N->getOperand(0));
+  SDValue V = DAG.getNode(N->getOpcode(), DL, NVT, ExtAcc, N->getOperand(1),
+                          N->getOperand(2));
+  return V;
```
Done

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11891,6 +11891,58 @@ SDValue TargetLowering::expandVECTOR_COMPRESS(SDNode *Node,
+  SDValue Acc = N->getOperand(0);
+  SDValue Input1 = N->getOperand(1);
+  SDValue Input2 = N->getOperand(2);
+  
```
nit: unnecessary newline.

JamesChesterman replied:
Done

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11891,6 +11891,58 @@ SDValue TargetLowering::expandVECTOR_COMPRESS(SDNode *Node,
+SDValue TargetLowering::expandPartialReduceMLA(SDNode *N, SelectionDAG &DAG) const {
+  SDLoc DL(N);
+  SDValue Acc = N->getOperand(0);
+  SDValue Input1 = N->getOperand(1);
+  SDValue Input2 = N->getOperand(2);
```
nit: what about `MulLHS` and `MulRHS` ?

JamesChesterman replied:
Done

sdesmalen-arm reviewed (approved):
LGTM!

JamesChesterman committed:
Rename variables and ensure conformity to coding standards.

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8118,15 +8125,35 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    SDValue Input = getValue(I.getOperand(1));
+    EVT InputVT = Input.getValueType();
+
+    assert(AccVT.getVectorElementType() == InputVT.getVectorElementType() &&
+           "Expected operands to have the same vector element type!");
+    assert(InputVT.getVectorElementCount().getKnownMinValue() %
+                   AccVT.getVectorElementCount().getKnownMinValue() ==
+               0 &&
+           "Expected the element count of the Input operand to be a positive "
+           "integer multiple of the element count of the Accumulator operand!");
```
These asserts belong in `SelectionDAG::getNode()` so that all node creations get equal protection.

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -135,6 +135,13 @@ static cl::opt<unsigned> SwitchPeelThreshold(
              "switch statement. A value greater than 100 will void this "
              "optimization"));
 
+// FIXME : This is a temporary flag, and is used to help transition to
+// performing lowering the proper way using the new PARTIAL_REDUCE_MLA ISD
+// nodes.
+static cl::opt<bool> NewPartialReduceLowering(
+    "new-partial-reduce-lowering", cl::init(false), cl::ReallyHidden,
+    cl::desc("Use the new method of lowering partial reductions."));
+
```
Is this really necessary?  To me it looks like we always end up using the same code the only difference being the flag trigger expansion earlier.  Is there a genuine problem the new command line option stops us from hitting?

I'm not strictly against having such a temporary flag but given the intrinsic is already tightly controlled it just feels overly paranoid to me.

JamesChesterman replied:
I think it is necessary? The command line flag will set I to a `ISD::PARTIAL_REDUCE_MLA` node, meaning it doesn't go to `visitTargetIntrinsic`. I think it will let us test the different DAG combines that are going to be made further down the line.

paulwalker-arm replied:
Perhaps that means the commend line flag is in the wrong place and we really just need something target specific that's used inside the AArch64 version of shouldExpandPartialReductionIntrinsic?

The problem I have with the current code is that within `SelectionDAGBuilder` we now have three possible outputs where I think only two is necessary:
1. `visitTargetIntrinsic` is called.
2. ISD::PARTIAL_REDUCE_UMLA is created and then later expanded during legalisation.

sdesmalen-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2881,6 +2890,13 @@ SDValue DAGTypeLegalizer::PromoteIntOp_VECTOR_FIND_LAST_ACTIVE(SDNode *N,
   return SDValue(DAG.UpdateNodeOperands(N, NewOps), 0);
 }
 
+SDValue DAGTypeLegalizer::PromoteIntOp_PARTIAL_REDUCE_MLA(SDNode *N,
+                                                          unsigned OpNo) {
+  SmallVector<SDValue, 1> NewOps(N->ops());
+  NewOps[OpNo] = GetPromotedInteger(N->getOperand(OpNo));
+  return SDValue(DAG.UpdateNodeOperands(N, NewOps), 0);
+}
+
```
I actually forgot to leave a comment about that, because I added an assert to test that OpNo > 1 when I checked out this patch locally, and found no failures, so OpNo = 0 should indeed be promoted as part of the result lowering.

Are you suggesting both operands need updating at the same time? (this function is otherwise called for both OpNo=1 and OpNo=2)

paulwalker-arm replied:
Yes, much like how `PromoteIntRes_PARTIAL_REDUCE_MLA` promotes the first operand to ensure the node remains consistent I think if one multiply operand needs promoting then by definition so does the other and thus both should be promoted at the same time.

JamesChesterman replied:
Done

JamesChesterman committed:
Address comments. Includes making cli option target-specific

Make comment describing the node more broad.
Promote both inputs at the same time.
Move assert statements to the getNode() function.
Make the command line option target-specific.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeDAG.cpp:
```diff
@@ -1245,6 +1245,11 @@ void SelectionDAGLegalize::LegalizeOp(SDNode *Node) {
+  case ISD::PARTIAL_REDUCE_SMLA:
+    Action = TLI.getOperationAction(Node->getOpcode(),
+                                    Node->getOperand(0).getValueType());
+    break;
```
Removed this as it was unnecessary. I think it must have been used to make a previous commit in this PR work. Thank you for spotting it.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:
```diff
@@ -503,6 +503,8 @@ SDValue VectorLegalizer::LegalizeOp(SDValue Op) {
   case ISD::VECREDUCE_FMAXIMUM:
   case ISD::VECREDUCE_FMINIMUM:
+  case ISD::PARTIAL_REDUCE_UMLA:
+  case ISD::PARTIAL_REDUCE_SMLA:
```
Done

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1571,6 +1581,11 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
       setOperationAction(ISD::MSTORE, VT, Custom);
     }
 
+    for (MVT VT : MVT::scalable_vector_valuetypes()) {
+      setOperationAction(ISD::PARTIAL_REDUCE_UMLA, VT, Expand);
+      setOperationAction(ISD::PARTIAL_REDUCE_SMLA, VT, Expand);
+    }
```
Given this new node is effectively opt-in I suggest adding an entry to `TargetLoweringBase::initActions()` so the operation is marked as expand for all value types.  Then in your follow-on patches you can selectively make some of them legal.

JamesChesterman replied:
Done

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -1358,6 +1365,9 @@ AArch64TargetLowering::AArch64TargetLowering(const TargetMachine &TM,
       setOperationAction(ISD::BSWAP, VT, Expand);
       setOperationAction(ISD::CTTZ, VT, Expand);
 
+      setOperationAction(ISD::PARTIAL_REDUCE_UMLA, VT, Expand);
+      setOperationAction(ISD::PARTIAL_REDUCE_SMLA, VT, Expand);
```
See comment attached to the scalable vector version of this code.

JamesChesterman replied:
Done

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -22010,8 +22027,13 @@ static SDValue performIntrinsicCombine(SDNode *N,
+    SDValue PRVal = DAG.getNode(ISD::PARTIAL_REDUCE_UMLA, DL,
+                                N->getValueType(0), N->getOperand(1), Input,
+                                DAG.getConstant(1, DL, Input.getValueType()));
+    return TLI.expandPartialReduceMLA(PRVal.getNode(), DAG);
```
Is this necessary? I'd expect to be able to return the `PARTIAL_REDUCE_UMLA` result and then let legalisation expand it later.

JamesChesterman replied:
Done, changed it so just the node is returned.

JamesChesterman committed:
Address comments. Includes improving assert statemetns.

Improve assert statements.
Remove unnecessary variable creation.
Change how operation actions are set for the nodes.

JamesChesterman commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1451,6 +1451,19 @@ enum NodeType {
   VECREDUCE_UMIN,
 
+  // PARTIAL_REDUCE_[U|S]MLA(Accumulator, Input1, Input2)
+  // Input1 and Input2 are multiplied together. This result is concatenated to
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:
```diff
@@ -1195,6 +1197,12 @@ void VectorLegalizer::Expand(SDNode *Node, SmallVectorImpl<SDValue> &Results) {
     return;
+  case ISD::PARTIAL_REDUCE_UMLA:
+  case ISD::PARTIAL_REDUCE_SMLA: {
+    SDLoc DL(Node);
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -1373,6 +1373,9 @@ void DAGTypeLegalizer::SplitVectorResult(SDNode *N, unsigned ResNo) {
     break;
+  case ISD::PARTIAL_REDUCE_UMLA:
+  case ISD::PARTIAL_REDUCE_SMLA:
+    SplitVecRes_PARTIAL_REDUCE_MLA(N);
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -3182,6 +3185,12 @@ void DAGTypeLegalizer::SplitVecRes_VP_REVERSE(SDNode *N, SDValue &Lo,
 }
 
+void DAGTypeLegalizer::SplitVecRes_PARTIAL_REDUCE_MLA(SDNode *N) {
+  SDLoc DL(N);
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -3381,6 +3390,9 @@ bool DAGTypeLegalizer::SplitVectorOperand(SDNode *N, unsigned OpNo) {
     break;
+  case ISD::PARTIAL_REDUCE_UMLA:
+  case ISD::PARTIAL_REDUCE_SMLA:
+    Res = SplitVecOp_PARTIAL_REDUCE_MLA(N);
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4435,6 +4447,13 @@ SDValue DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM(SDNode *N) {
 }
 
+SDValue DAGTypeLegalizer::SplitVecOp_PARTIAL_REDUCE_MLA(SDNode *N) {
+  SDLoc DL(N);
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7883,6 +7854,30 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+  case ISD::PARTIAL_REDUCE_SMLA: {
+    EVT AccVT = N1.getValueType();
+    EVT Input1VT = N2.getValueType();
+    EVT Input2VT = N3.getValueType();
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7883,6 +7854,30 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+    EVT AccVT = N1.getValueType();
+    EVT Input1VT = N2.getValueType();
+    EVT Input2VT = N3.getValueType();
+    assert(Input1VT == Input2VT &&
```
Done (used `isVector()`).

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7883,6 +7854,30 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+    assert(Input1VT == Input2VT &&
+           "Expected the second and third operands of the PARTIAL_REDUCE_MLA "
+           "node to have the same type!");
+    assert(VT == AccVT &&
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7883,6 +7854,30 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+           "the same type as its result!");
+    assert(Input1VT.getVectorElementCount().getKnownMinValue() %
+                   AccVT.getVectorElementCount().getKnownMinValue() ==
+               0 &&
```
Yes, thank you. Done.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -7883,6 +7854,30 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
+           "PARTIAL_REDUCE_MLA node to be a positive integer multiple of the "
+           "element count of the first operand and result!");
+    assert(Input1VT.getVectorElementType().getSizeInBits() <=
+               AccVT.getVectorElementType().getSizeInBits() &&
```
Done (needed to be done on the operands not the types though).

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -8118,15 +8118,17 @@ void SelectionDAGBuilder::visitIntrinsicCall(const CallInst &I,
+    SDValue Acc = getValue(I.getOperand(0));
+    EVT AccVT = Acc.getValueType();
+    SDValue Input = getValue(I.getOperand(1));
+    EVT InputVT = Input.getValueType();
```
Done (also removed VT variables).

JamesChesterman committed:
Address comments, includes fix for using wrong signedness

Fix wrong signedness in promotion function.
Make expand code depend on node signedness not on operand extend.

JamesChesterman committed:
Fix issue with expand function.

Was seeing if the operand opcodes were PARTIAL_REDUCE_MLA nodes.
Now looking at their parent.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -2881,6 +2890,13 @@ SDValue DAGTypeLegalizer::PromoteIntOp_VECTOR_FIND_LAST_ACTIVE(SDNode *N,
+SDValue DAGTypeLegalizer::PromoteIntOp_PARTIAL_REDUCE_MLA(SDNode *N) {
+  SmallVector<SDValue, 1> NewOps(N->ops());
+  NewOps[1] = GetPromotedInteger(N->getOperand(1));
+  NewOps[2] = GetPromotedInteger(N->getOperand(2));
```
Done

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/LegalizeIntegerTypes.cpp:
```diff
@@ -6196,6 +6212,15 @@ SDValue DAGTypeLegalizer::PromoteIntRes_VECTOR_FIND_LAST_ACTIVE(SDNode *N) {
+  SDLoc DL(N);
+  EVT VT = N->getValueType(0);
+  EVT NVT = TLI.getTypeToTransformTo(*DAG.getContext(), VT);
+  SDValue ExtAcc = GetPromotedInteger(N->getOperand(0));
```
OK, thank you for explaining

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11890,6 +11891,61 @@ SDValue TargetLowering::expandVECTOR_COMPRESS(SDNode *Node,
+
+  auto ExtendToAccEltVT = [&](SDValue V) {
+    unsigned ExtOpc = V->getOpcode() == ISD::SIGN_EXTEND ? ISD::SIGN_EXTEND
+                                                         : ISD::ZERO_EXTEND;
```
Done. Changed it so both operands are promoted with the same signedness, and it only depends on `N`'s opcode.

JamesChesterman commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -22010,8 +22019,11 @@ static SDValue performIntrinsicCombine(SDNode *N,
       return WideAdd;
-    return DAG.getPartialReduceAdd(SDLoc(N), N->getValueType(0),
-                                   N->getOperand(1), N->getOperand(2));
+    SDLoc DL(N);
+    SDValue Input = N->getOperand(2);
+    return DAG.getNode(ISD::PARTIAL_REDUCE_UMLA, DL, N->getValueType(0),
+                       N->getOperand(1), Input,
+                       DAG.getConstant(1, DL, Input.getValueType()));
```
I've kept this the same in the end, just because the `DL` and `Input` are used in the `DAG.getConstant` call within this too, unless there's something I'm missing?

paulwalker-arm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -4435,6 +4448,12 @@ SDValue DAGTypeLegalizer::SplitVecOp_VECTOR_HISTOGRAM(SDNode *N) {
 }
 
+SDValue DAGTypeLegalizer::SplitVecOp_PARTIAL_REDUCE_MLA(SDNode *N) {
+  SDValue Res = TLI.expandPartialReduceMLA(N, DAG);
+  ReplaceValueWith(SDValue(N, 0), Res);
+  return SDValue();
```
Does just `return TLI.expandPartialReduceMLA(N, DAG)` work here?

JamesChesterman replied:
Yes, changed it now.

paulwalker-arm reviewed (approved):
A few final recommendations but otherwise this looks good to me.  I'll admit to not be convinced about the implementation of `SplitVecRes_PARTIAL_REDUCE_MLA` but figure that'll be short lived and so if you're sure it works then that's good enough for me.

preames commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -3182,6 +3186,11 @@ void DAGTypeLegalizer::SplitVecRes_VP_REVERSE(SDNode *N, SDValue &Lo,
 
+void DAGTypeLegalizer::SplitVecRes_PARTIAL_REDUCE_MLA(SDNode *N) {
+  SDValue Res = TLI.expandPartialReduceMLA(N, DAG);
+  ReplaceValueWith(SDValue(N, 0), Res);
```
Based on the precedent of SplitVecRes_VECTOR_SPLICE - which is the only example of expanding during SplitVecRes I found - I think you need to manually split the result of the expand into Lo and Hi here.

paulwalker-arm replied:
Oh nice, thanks for finding this.

JamesChesterman replied:
Doing it this way did not change the result of any tests in the end. But I have changed it to doing it this way to make it more similar to what is already done. Thank you for pointing it out.

paulwalker-arm commented on llvm/include/llvm/CodeGen/ISDOpcodes.h:
```diff
@@ -1451,6 +1451,20 @@ enum NodeType {
+  // The partial reduction nodes sign or zero extend Input1 and Input2 to the
+  // element type of Accumulator before multiplying their results.
+  // This result is concatenated to the Accumulator, and this is then reduced,
+  // using addition, to the result type.
```
The intrinsic and matching ISD node are deliberate in not defining the order in which the elements are reduced because their output is only expected to feed into themselves or an equivalent vector.reduce operation and thus the ordering does not matter.  Essentially, add reductions can now be represented in LLVM as a two phase operation (one in-loop and the other out-of-loop) in the most relaxed way possible.

preames replied:
Then you need to say that clearly in the specification text.  At the moment, it looks like merely an unstated assumption, not a specifically reserved behavior.  The bit about the result only being valid for use by further reductions (i.e. non-order preserving operations) is really important to clearly say.

Seriously, I read your specification and didn't not understand this.  I doubt I'm the only one.

JamesChesterman replied:
I've added a statement to try and make it more specific now. Thank you.

JamesChesterman committed:
Change splitting functions.

Adjust ISDOpcode description.
Rename variables in expand function.
Remove unnecessary assert statement.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11890,6 +11891,62 @@ SDValue TargetLowering::expandVECTOR_COMPRESS(SDNode *Node,
+  unsigned ExtOpc = N->getOpcode() == ISD::PARTIAL_REDUCE_SMLA
+                        ? ISD::SIGN_EXTEND
+                        : ISD::ZERO_EXTEND;
+  EVT MulLHSVT = MulLHS.getValueType();
```
Done. Removed `MulLHSVT` and renamed `FullTy` to `MulOpVT`. I've then renamed `ReducedTy` to `AccVT` to make it more similar to the rest of the variable names.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11890,6 +11891,62 @@ SDValue TargetLowering::expandVECTOR_COMPRESS(SDNode *Node,
+  EVT MulLHSVT = MulLHS.getValueType();
+  assert(MulLHSVT == MulRHS.getValueType() &&
+         "The second and third operands of a PARTIAL_REDUCE_MLA node must have "
+         "the same value type!");
```
Done, removed assert statement.

JamesChesterman commented on llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:
```diff
@@ -11890,6 +11891,62 @@ SDValue TargetLowering::expandVECTOR_COMPRESS(SDNode *Node,
+         "The second and third operands of a PARTIAL_REDUCE_MLA node must have "
+         "the same value type!");
+  EVT ExtVT = MulLHSVT.changeVectorElementType(
+      Acc.getValueType().getVectorElementType());
```
Done, removed `ExtVT` initialisation and renamed `NewVT` to `ExtMulOpVT`.
